### ISSUE TITLE
Rule for spaces around "=" character

### ISSUE THREAD

**Issue by SamuelBolduc** [eslint/eslint#3299]
I looked through the available rules but I can't seem to find how to enforce having spaces around a `=` character.

This should give a warning:
`const my_object={key: 'value'};`

This should not give a warning and should be considered valid:
 `const my_object = {key: 'value'};`

Is there already a rule to enforce this?

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
I don't believe so. We should definitely have one.

**Comment by IanVS** [comment]
What should it be called?

**Comment by BYK** [comment]
I expected `space-infix-ops` to cover this but obviously `=` is not an infix operator. How about `space-assigment` as the name? This can cover the `:` in object expressions.

**Comment by IanVS** [comment]
For `:`, there is already [key-spacing](http://eslint.org/docs/rules/key-spacing), does that cover what you had in mind?

**Comment by BYK** [comment]
>For :, there is already key-spacing, does that cover what you had in mind?

Yeah, I realized this half an hour I made this comment :) Still no solution for `=` case though.

**Comment by michaelficarra** [comment]
@BYK `space-assignment` would be an inappropriate name, since this is not assignment and assignment already gets checked by the `space-infix-ops` rule. I suggest `space-var-init`.

**Comment by IanVS** [comment]
I presume this would apply to `const` and `let` as well, so perhaps `space-variable-init` to avoid confusion?

**Comment by nzakas** [comment]
:+1:

**Comment by gyandeeps** [comment]
I think the issue started with `=` character then some how the convo went into the use of `=` inside variable declarations. 
Is our aim to create this rule for `=` inside variables only or everywhere?
CC @SamuelBolduc

**Comment by michaelficarra** [comment]
@gyandeeps Just variable declarations. Otherwise it would overlap with `space-infix-ops`.

**Comment by IanVS** [comment]
As @michaelficarra mentioned, `space-infix-ops` covers assignment. Are there other uses for = besides initialization and assignment?

**Comment by mysticatea** [comment]
@IanVS default values.

```js
function foo(a = 0) { }

var {a = 0} = bar;
```

**Comment by BYK** [comment]
I think we are lost in weeds. `space-infix-ops` supports this. PR upcoming for tests and updated documentation.

**Comment by BYK** [comment]
>but obviously = is not an infix operator.

lol at myself

**Comment by michaelficarra** [comment]
@BYK `space-infix-ops` should not handle the variable initialiser or default value cases.

**Comment by BYK** [comment]
@michaelficarra - why wouldn't it? Also it already does, that's my point.

**Comment by BYK** [comment]
>Also it already does, that's my point.

Except for the default value case. It already does handle variable initializer, including `const` and `let`.

**Comment by michaelficarra** [comment]
Because the `op` in `space-infix-ops` stands for operator, and that is not an operator.

**Comment by BYK** [comment]
@michaelficarra - I understand the semantics and if we prefer to be pure you are correct. But I, like some other people assumed it would cover `=` too. Even weirder, looks like you were the one who added this behavior to fix #464 2 years ago so I'm really confused about the change of heart now.

**Comment by IanVS** [comment]
@ilyavolodin did you mean to close this issue from https://github.com/eslint/eslint/commit/ba818eee0fd69a50abe6289ae27b7ae4248c733d?  It seems like maybe that should have been a `(refs #2399)`.

**Comment by BYK** [comment]
@IanVS - I think it was automatic and it kind of addresses the original issue since `space-infix-ops` covers that case already.
